File: src/main/java/com/twitter/route/Route.java
package com.twitter.route;

/**
 * Created by mariusz on 18.07.16.
 */
public final class Route {

    private static final String API = "/api";
    private static final String ALL = "/**";
    private static final String SIZE = "/{size}";
    private static final String PAGE = "/{page}";
    private static final String USER = "/user";
    private static final String USERNAME = "/{username}";
    private static final String USER_ID = "/{userId}";
    private static final String TWEET = "/tweet";
    private static final String TWEET_ID = "/{tweetId}";
    private static final String COMMENT = "/comment";
    private static final String COMMENT_ID = "/{commentId}";
    private static final String FOLLOW = "/follow";
    private static final String FOLLOWING = "/following";
    private static final String FOLLOWERS = "/followers";
    private static final String COUNT = "/count";
    private static final String VERIFY = "/verify";
    private static final String LATEST = "/latest";
    private static final String OLDEST = "/oldest";
    private static final String POPULAR = "/popular";
    private static final String VOTE = "/vote";
    private static final String VOTE_ID = "/{voteId}";
    private static final String REPORT = "/report";
    private static final String REPORT_ID = "/{reportId}";
    private static final String STATUS = "/status";
    private static final String CATEGORY = "/category";
    private static final String REPORT_STATUS = "/{reportStatus}";
    private static final String REPORT_CATEGORY = "/{reportCategory}";
    private static final String AVATAR = "/avatar";
    private static final String FAVOURITES = "/favourites";
    private static final String TAGS = "/tags";
    private static final String HOURS = "/{hours}";
    private static final String PASSWORD = "/password";
    private static final String ROLE = "/role";
    private static final String EMAIL = "/email";
    private static final String NOTIFICATION = "/notification";
    private static final String NOTIFICATION_ID = "/{notificationId}";
    private static final String SEEN = "/{seen}";
    private static final String QUERY = "/query";


    public static final String LOGIN_URL = API + "/login";
    public static final String LOGOUT_URL = API + "/logout";
    public static final String ALL_REQUESTS_REGEX = API + ALL;
    public static final String VERIFY_USER_URL = API + USER + VERIFY + "/{verifyKey}";
    public static final String VERIFY_USER_URL_REGEX = API + USER + VERIFY + ALL;

    public static final String USER_URL = API + USER;
    public static final String USER_BY_ID = USER_URL + USER_ID;
    public static final String USER_PASSWORD_CHANGE = USER_URL + USER_ID + PASSWORD;
    public static final String USER_ROLE_CHANGE = USER_URL + USER_ID + ROLE;
    public static final String USER_EMAIL_CHANGE = USER_BY_ID + EMAIL;
    public static final String USER_FOLLOW = USER_URL + FOLLOW + USER_ID;
    public static final String USER_GET_ALL = USER_URL + PAGE + SIZE;
    public static final String USER_COUNT_GET_ALL = USER_URL + COUNT;
    public static final String REGISTER_USER = USER_URL;
    public static final String USER_WITH_USERNAME_LIKE = USER_URL + QUERY + USERNAME + PAGE + SIZE;
    public static final String USER_GET_FOLLOWERS = USER_BY_ID + FOLLOWERS + PAGE + SIZE;
    public static final String USER_COUNT_GET_FOLLOWERS = USER_BY_ID + FOLLOWERS + COUNT;
    public static final String USER_GET_FOLLOWING = USER_BY_ID + FOLLOWING + PAGE + SIZE;
    public static final String USER_COUNT_GET_FOLLOWING = USER_BY_ID + FOLLOWING + COUNT;
    public static final String USER_AVATAR = USER_BY_ID + AVATAR;
    public static final String USER_FAVOURITE_TAGS = USER_BY_ID + TAGS;

    public static final String TAG_WITH_TEXT_LIKE = API + TAGS + QUERY + "/{tagText}" + PAGE + SIZE;

    public static final String TWEET_URL = API + TWEET;
    public static final String TWEET_BY_ID = TWEET_URL + TWEET_ID;
    public static final String TWEET_GET_ALL = TWEET_URL + PAGE + SIZE;
    public static final String TWEETS_FROM_FOLLOWINGS_USERS = TWEET_URL + "/my" + USER_ID + PAGE + SIZE;
    public static final String TWEETS_FROM_USER = TWEET_URL + USER + USER_ID + PAGE + SIZE;
    public static final String TWEETS_MOST_VOTED = TWEET_URL + POPULAR + HOURS + PAGE + SIZE;
    public static final String TWEET_VOTE = TWEET_URL + VOTE;
    public static final String TWEET_USER_VOTE = TWEET_BY_ID + VOTE;
    public static final String TWEET_USER_VOTE_COUNT = TWEET_BY_ID + COUNT + VOTE + VOTE_ID;
    public static final String TWEETS_WITH_TAGS = TWEET_URL + TAGS + "/{tags}" + PAGE + SIZE;
    public static final String TWEETS_WITH_TAGS_POPULAR = TWEET_URL + TAGS + "/{tags}" + POPULAR + HOURS + PAGE + SIZE;
    public static final String TWEETS_FROM_USER_FAVOURITES = USER_BY_ID + FAVOURITES + PAGE + SIZE;
    public static final String TWEET_TO_USER_FAVOURITES = USER_URL + FAVOURITES + TWEET_ID;

    public static final String COMMENT_URL = API + COMMENT;
    public static final String COMMENTS_FROM_TWEET = TWEET_BY_ID + COMMENT + PAGE + SIZE;
    public static final String COMMENT_BY_ID = COMMENT_URL + COMMENT_ID;
    public static final String COMMENTS_FROM_USER = COMMENT_URL + USER + USER_ID + PAGE + SIZE;
    public static final String COMMENTS_LATEST = TWEET_BY_ID + COMMENT + LATEST + PAGE + SIZE;
    public static final String COMMENTS_OLDEST = TWEET_BY_ID + COMMENT + OLDEST + PAGE + SIZE;
    public static final String COMMENTS_POPULAR = TWEET_BY_ID + COMMENT + POPULAR + PAGE + SIZE;
    public static final String COMMENT_VOTE = COMMENT_URL + VOTE;
    public static final String COMMENT_VOTE_BY_COMMENT_ID = COMMENT_BY_ID + VOTE;
    public static final String COMMENT_USER_VOTE_COUNT = COMMENT_BY_ID + COUNT + VOTE + VOTE_ID;


    public static final String REPORT_URL = API + REPORT;
    public static final String REPORT_GET_LATEST = REPORT_URL + LATEST + PAGE + SIZE;
    public static final String REPORT_GET_FROM_USER = REPORT_URL + PAGE + SIZE;
    public static final String REPORT_BY_ID = REPORT_URL + REPORT_ID;
    public static final String REPORT_GET_ALL_BY_STATUS = REPORT_URL + STATUS + REPORT_STATUS + PAGE + SIZE;
    public static final String REPORT_GET_ALL_BY_CATEGORY = REPORT_URL + CATEGORY + REPORT_CATEGORY + PAGE + SIZE;
    public static final String REPORT_GET_ALL_BY_STATUS_AND_CATEGORY = REPORT_URL + STATUS + REPORT_STATUS +
            CATEGORY + REPORT_CATEGORY + PAGE + SIZE;

    public static final String NOTIFICATION_URL = API + NOTIFICATION;
    public static final String NOTIFICATION_BY_ID = NOTIFICATION_URL + NOTIFICATION_ID;
    public static final String NOTIFICATION_GET = NOTIFICATION_URL + SEEN + PAGE + SIZE;

}

File: src/main/resources/static/src/domain/route.ts
import Tag = Models.Tag;
/**
 * Created by mariusz on 30.08.16.
 */

export const BASE_URL = `http://localhost:8080`;
export const USER_URL = `/api/user`;
export const REGISTER = USER_URL;
export const CURRENT_USER = USER_URL;
export const LOGIN = `/api/login`;
export const TWEET_URL = `/api/tweet`;
export const COMMENT_URL = `/api/comment`;
export const REPORT_URL = `/api/report`;
export const TAG_URL = `/api/tweet/tags`;
export const USER_BY_ID = (userId:number) => `${USER_URL}/${userId}`;
export const USER_CHANGE_PASSWORD = (userId:number) => `${USER_BY_ID(userId)}/password`;
export const USER_CHANGE_ROLE = (userId:number) => `${USER_BY_ID(userId)}/role`;
export const USER_CHANGE_EMAIL = (userId:number) => `${USER_BY_ID(userId)}/email`;
export const USER_FOLLOWERS_URL = (userId:number) => `${USER_BY_ID(userId)}/followers`;
export const USER_FOLLOWING_URL = (userId:number) => `${USER_BY_ID(userId)}/following`;
export const COMMENT_BY_ID = (commentId:number) => `${COMMENT_URL}/${commentId}`;
export const TWEET_BY_ID = (tweetId:number) => `${TWEET_URL}/${tweetId}`;
export const TWEET_GET_ALL = (page:number, size:number) => `${TWEET_URL}/${page}/${size}`;
export const TWEET_VOTE_GET_BY_ID = (tweetId:number) => `${TWEET_BY_ID(tweetId)}/vote`;
export const TWEET_VOTE = `${TWEET_URL}/vote`;
export const COMMENT_VOTE = `${COMMENT_URL}/vote`;
export const CHANGE_USER_AVATAR = (userId:number) => `${USER_BY_ID(userId)}/avatar`;
export const TWEET_FAVOURITE = (tweetId:number | string) =>`${USER_URL}/favourites/${tweetId}`;
export const USER_FAVOURITE_TWEETS = (userId:number, page:number, size:number) => `${USER_URL}/${userId}/favourites/${page}/${size}`;
export const COMMENTS_FROM_TWEET = (tweetId:number, page:number, size:number) => `${TWEET_BY_ID(tweetId)}/comment/${page}/${size}`;
export const COMMENT_VOTE_BY_ID = (commentId:number) => `${COMMENT_BY_ID(commentId)}/vote`;
export const TWEETS_FROM_USER = (userId:number, page:number, size:number) => `${TWEET_URL}/user/${userId}/${page}/${size}`;
export const TWEETS_FROM_FOLLOWING_USERS = (userId:number, page:number, size:number) => `${TWEET_URL}/my/${userId}/${page}/${size}`;
export const FOLLOW_USER = (userId:number) =>`${USER_URL}/follow/${userId}`;
export const USER_FOLLOWERS = (userId:number, page:number, size:number) => `${USER_FOLLOWERS_URL(userId)}/${page}/${size}`;
export const USER_FOLLOWING = (userId:number, page:number, size:number) => `${USER_FOLLOWING_URL(userId)}/${page}/${size}`;
export const USER_FOLLOWERS_COUNT = (userId:number) => `${USER_FOLLOWERS_URL(userId)}/count`;
export const USER_FOLLOWING_COUNT = (userId:number) => `${USER_FOLLOWING_URL(userId)}/count`;
export const USER_REPORTS = (page:number, size:number) => `${REPORT_URL}/${page}/${size}`;
export const REPORTS_LATEST = (page:number, size:number) => `${REPORT_URL}/latest/${page}/${size}`;
export const REPORTS_CATEGORY = (category:string, page:number, size:number) => `${REPORT_URL}/category/${category}/${page}/${size}`;
export const REPORTS_STATUS = (status:string, page:number, size:number) => `${REPORT_URL}/status/${status}/${page}/${size}`;
export const REPORTS_BY_STATUS_AND_CATEGORY = (status:string, category:string, page:number, size:number) => `${REPORT_URL}/status/${status}/category/${category}/${page}/${size}`;
export const TWEETS_BY_TAGS = (tags:string[], page:number, size:number) => `${TAG_URL}/${tags.join(',')}/${page}/${size}`;
export const TWEETS_MOST_VOTED = (hours:number, page:number, size:number) => `${TWEET_URL}/popular/${hours}/${page}/${size}`;
export const USER_FAVOURITE_TAGS = (userId:number) => `${USER_BY_ID(userId)}/tags`;
export const TWEET_VOTE_COUNT = (tweetId:number, vote:string) => `${TWEET_BY_ID(tweetId)}/count/vote/${vote}`;
export const COMMENT_VOTE_COUNT = (commentId:number, vote:string) => `${COMMENT_BY_ID(commentId)}/count/vote/${vote}`;